New Workflow

XedinUnknown opened the pull request:
- Added Composer
- Mollie API is now a dependency, not a submodule.
- Mollie classes now loaded by Composer according to PSR-0.
- Removed non-project assets.
- Moved relevant things to project root, removing the separation between project and plugin.
- Added Docker Compose configuration for testing and development.
- WordPress now a correctly declared dependency.

Addresses [MOL-19][].

[MOL-19]: https://inpsyde.atlassian.net/browse/MOL-19

XedinUnknown commented:
Oh! Hmm. @Dinamiko, what happens before this error? Did it install the dependencies correctly?

Dinamiko commented:
@XedinUnknown dependences installed correctly, error happens just after `Generating autoload files` message.

widoz commented on .env:
```diff
@@ -0,0 +1,101 @@
+
+# Linux (uid 1000 gid 1000)
+
+PHP_TAG=7.3-dev-4.12.12
```
Would not better to enable php 5.6 since the plugin must be compatible with that version?

XedinUnknown replied:
This is counter-intuitive, but if you look in the override file, you will see that it uses the WordPress image, not the PHP one. This is currently the last configuration that worked, and with it you cannot switch the PHP version; only WP version. However, the dependencies are installed with PHP 5.6.40.

widoz commented on composer.json:
```diff
@@ -1,25 +1,41 @@
     "require-dev": {
-        "phpunit/phpunit": "^5",
-        "brain/monkey": "^2.0@dev",
-        "ptrofimov/xpmock": "^1"
```
Why those packages are not included? We need them for testing

"phpunit/phpunit": "^5",
"brain/monkey": "^2.0@dev",
"ptrofimov/xpmock": "^1"

XedinUnknown replied:
I could not have removed these, because when I started working on this task, there was no Composer configuration. Most likely, the later changes weren't merged correctly onto this branch.

widoz commented on composer.json:
```diff
@@ -1,25 +1,41 @@
-    "autoload-dev": {
-        "psr-4": {
-            "Mollie\\WooCommerce\\Tests\\": "tests/php",
-            "Mollie\\WooCommerce\\Tests\\Unit\\": "tests/php/Unit"
```
We need them to load the tests classes

XedinUnknown replied:
See above.

widoz commented on .env:
```diff
@@ -0,0 +1,101 @@
+
+### --- VANILLA WORDPRESS ----
+
+WORDPRESS_TAG=5-4.16.1
```
I would use 3.8 version until we decide what to do.

XedinUnknown replied:
Given that right now we can only test against one WP version, I believe that latest stable is more beneficial. More people have WP 5.2 than any other version, according to [official statistics][1]. Once we can test against multiple versions (not before we have the integration with Codeship), I am all for including 3.8.

[1]: https://wordpress.org/about/stats/#wp_versions

widoz commented on docker.mk:
```diff
@@ -0,0 +1,62 @@
+
+## up	:	Start up containers.
+up:
+	@echo "Starting up containers for for $(PROJECT_NAME)..."
```
I don't know if this is a typo but **for** appear twice.

XedinUnknown replied:
Now tracked in wodby/docker4wordpress#50.

widoz commented on mollie-payments-for-woocommerce.php:
```diff
@@ -0,0 +1,197 @@
+ * Plugin Name: Mollie Payments for WooCommerce
+ * Plugin URI: https://www.mollie.com
+ * Description: Accept payments in WooCommerce with the official Mollie plugin
+ * Version: 5.1.8
```
The version should be 5.2.1

XedinUnknown replied:
Here it says "5.1.8" most likely because that was the version from which I branched. If it still says that, then later changes weren't merged correctly into this branch.

widoz commented on mollie-payments-for-woocommerce.php:
```diff
@@ -0,0 +1,197 @@
+ * Author: Mollie
+ * Author URI: https://www.mollie.com
+ * Requires at least: 3.8
+ * Tested up to: 5.1
```
Tested up to : 5.2

XedinUnknown replied:
See above. Also, this is probably something that should be addressed in a `release/*` branch.

widoz commented on src/Mollie/Wc/Exception/IncompatiblePlatform.php:
```diff
@@ -0,0 +1,50 @@
+    const API_CLIENT_NOT_COMPATIBLE   = 2000;
+}
+
+
```
Code below should not be here

XedinUnknown replied:
Totally :+1:.

XedinUnknown commented:
@widoz,

1. a136bb3f doesn't have the `mollie-payments-for-woocommerce`; it only comes back after the changes you added.
2. `readme.txt` is a concern for the WordPress.org repository, as @danielhuesken said earlier. Will add the changelog file.

XedinUnknown commented:
@widoz, also, does the plugin work for you? Are you experiencing the symptoms described by @Dinamiko?

XedinUnknown commented:
I will remove the `.gitmodules` file, since we don't use modules anymore.

widoz commented:
> @XedinUnknown I´m getting this error when I run `composer install`:
> 
> ```
> Generating autoload files
> PHP Fatal error:  Uncaught Error: Class 'Automattic\Jetpack\Autoloader\AutoloadGenerator' not found in /Users/emiliocastells/Desktop/WooCommerce/vendor/automattic/jetpack-autoloader/src/CustomAutoloaderPlugin.php:84
> Stack trace:
> #0 [internal function]: Automattic\Jetpack\Autoloader\CustomAutoloaderPlugin->postAutoloadDump(Object(Composer\Script\Event))
> #1 phar:///usr/local/bin/composer/src/Composer/EventDispatcher/EventDispatcher.php(171): call_user_func(Array, Object(Composer\Script\Event))
> #2 phar:///usr/local/bin/composer/src/Composer/EventDispatcher/EventDispatcher.php(96): Composer\EventDispatcher\EventDispatcher->doDispatch(Object(Composer\Script\Event))
> #3 phar:///usr/local/bin/composer/src/Composer/Autoload/AutoloadGenerator.php(312): Composer\EventDispatcher\EventDispatcher->dispatchScript('post-autoload-d...', true, Array, Array)
> #4 phar:///usr/local/bin/composer/src/Composer/Installer.php(302): Composer\Autoload\AutoloadGenerator->dump(Object(Composer\Config), Object(Composer\Repository\InstalledFilesystemRepository in /Users/emiliocastells/Desktop/WooCommerce/vendor/automattic/jetpack-autoloader/src/CustomAutoloaderPlugin.php on line 84
> 
> Fatal error: Uncaught Error: Class 'Automattic\Jetpack\Autoloader\AutoloadGenerator' not found in /Users/emiliocastells/Desktop/WooCommerce/vendor/automattic/jetpack-autoloader/src/CustomAutoloaderPlugin.php:84
> Stack trace:
> #0 [internal function]: Automattic\Jetpack\Autoloader\CustomAutoloaderPlugin->postAutoloadDump(Object(Composer\Script\Event))
> #1 phar:///usr/local/bin/composer/src/Composer/EventDispatcher/EventDispatcher.php(171): call_user_func(Array, Object(Composer\Script\Event))
> #2 phar:///usr/local/bin/composer/src/Composer/EventDispatcher/EventDispatcher.php(96): Composer\EventDispatcher\EventDispatcher->doDispatch(Object(Composer\Script\Event))
> #3 phar:///usr/local/bin/composer/src/Composer/Autoload/AutoloadGenerator.php(312): Composer\EventDispatcher\EventDispatcher->dispatchScript('post-autoload-d...', true, Array, Array)
> #4 phar:///usr/local/bin/composer/src/Composer/Installer.php(302): Composer\Autoload\AutoloadGenerator->dump(Object(Composer\Config), Object(Composer\Repository\InstalledFilesystemRepository in /Users/emiliocastells/Desktop/WooCommerce/vendor/automattic/jetpack-autoloader/src/CustomAutoloaderPlugin.php on line 84
> ```

@Dinamiko @XedinUnknown I have the same problem when I run `composer install --no-dev --optimize-autoloader`

widoz commented:
@XedinUnknown Since we haven't decided what to do with `.idea` would make sense to remove it for now?

XedinUnknown commented:
Many things in the PHPStorm project configuration are the same for all developers. Examples include:

1. Vendor library paths, whether as PHPStorm components, or include/exclude paths, etc.
2. Other paths, such as source, test, asset directories.
3. VCS settings. We all use the same repo with the same settings, etc.
4. Other things like database access credentials. Now that we are using Docker environments, we can share more things than ever before.

Of course, some things are very specific to the developer and their environment. For example, the code inspection settings, workspace, editor colours, shortcuts, etc. Right now these things are excluded from the project.

It is my opinion that the project name should also be the same, since it's the same project. I would be OK with changing that per developer, but PHPStorm does not allow that, because the name of the `.iml` file appears to determine project name, and IMHO that file should be shared.

In any case, I am happy to discuss which files to include/exclude. One thing IMO is certain: setting up the project again every time is tedious and a waste of time.

widoz commented:
@XedinUnknown to speed up the process I would add the phpstorm configuration later on, since isn't mandatory/blocker because everyone can use their own favorite IDE or Editor waiting to solve that part will unnecessary block the entire work done in this branch.

widoz commented:
@XedinUnknown 

Regarding "That's because AFAIK this is the way PSR-0 works." Do we have plans to move to psr-4?

XedinUnknown commented:
About PHPStorm configuration, there are other people, like @danielhuesken, which would like to try the project. If we remove the configs, he will again have to configure it, etc. I see no harm in simply keeping the current configuration until decisions are made. It's better than nothing.

In relation to PSR-4, I have no plans to convert it. What I tried to do is to adapt the package with minimal code changes, and PSR-0 did not require refactoring of class names. But if there is a need and a decision to convert to PSR-4, then sure. It's just more complicated and prone to errors, not to mention that it's a major BC break.

XedinUnknown commented:
By the way, @danielhuesken reports the same symptoms as @Dinamiko when installing dependencies.

widoz commented:
@XedinUnknown Regarding the .idea inclusion, I lost a lot of customization phpstorm allow me to do such as debug, test framework, composer, servers etc... I think until we have better knowledge about which is included in the .idea and what should be included better to not include that directory.

XedinUnknown commented:
> > About PHPStorm configuration, there are other people, like @danielhuesken, which would like to try the project. If we remove the configs, he will again have to configure it, etc. I see no harm in simply keeping the current configuration until decisions are made. It's better than nothing.
> 
> 
> Which kind of configurations are necessary in order to try the project? I mean, it's not really necessary to test the project in a functional way, them are "useful" if you need to develop, am I wrong?

Correct. The configuration has been removed from here. This concern is now addressed in #306.

XedinUnknown commented on Makefile:
```diff
@@ -0,0 +1,8 @@
+
+.PHONY: test
+
+PHP_VER ?= 7.2
```
The Makefile includes stuff for building the images themselves. This is part of that. It sets the PHP version to be used in `run.sh`.

XedinUnknown commented:
For some reason, GitHub does not allow me to reply in the correct place to [this particular question][1], so I am replying here.

> Sorry I missed one thing, which version of PHP is running in WordPress Container?

The PHP version depends on the WordPress version with the current image used in this configuration. In this case, it's PHP 7.3, as it says in that Makefile. I can confirm this:

```
$ docker-compose run php php -v
sudo -E init
PHP 7.3.6 (cli) (built: May 31 2019 02:54:22) ( NTS )
Copyright (c) 1997-2018 The PHP Group
Zend Engine v3.3.6, Copyright (c) 1998-2018 Zend Technologies
    with Zend OPcache v7.3.6, Copyright (c) 1999-2018, by Zend Technologies

```

[1]: https://github.com/mollie/WooCommerce/pull/303#discussion_r314654152

XedinUnknown commented on .idea/modules.xml:
```diff
@@ -0,0 +1,8 @@
+<project version="4">
+  <component name="ProjectModuleManager">
+    <modules>
+      <module fileurl="file://$PROJECT_DIR$/.idea/WooCommerce.iml" filepath="$PROJECT_DIR$/.idea/WooCommerce.iml" />
```
No more configuration in this PR.

widoz commented:
@XedinUnknown I'll be glad to approve the pr, just could you resolve the conflicts? Let me know if you need help on that.

widoz commented:
@XedinUnknown I resolved the conflicts, please have a look.

Ps. I still see the custom `autoload.php` file, should be removed?

XedinUnknown commented:
@widoz, thanks very much for resolving conflicts! I've removed the proprietary autoloader, and also added some helpful changes. Please take a look, and approve if you're happy.

widoz commented:
@XedinUnknown I fixed the conflicts again, hope this is the last time shouldn't be any other PR for now but would be nice if we can merge this in develop as soon as possible.

Please check my changes and let me know about the `.idea´ directory which is in the PR.

danielhuesken commented:
The version we schould release it is 5.4.0 ther no bracking changes so that we need a version 6.

XedinUnknown commented:
@widoz, I fixed the `.idea` directory, updated Docker setup, fixed a bug, etc. Please check and approve if possible. Now that we have the go-ahead to release this, we should merge ASAP.
